internal/codegen/tpl_stencil_arg.go

```diff
@@ -0,0 +1,222 @@
+	}
+	arg := mf.Arguments[pth]
+
+	// If there's a "from" we should handle that now before anything else,
+	// so that its definition is used.
+	if arg.From != "" {
+		fromArg, err := s.resolveFrom(ctx, pth, &arg)
+		if err != nil {
+			return "", err
+		}
+		// Guaranteed to not be nil
+		arg = *fromArg
+	}
+
+	mapInf := make(map[interface{}]interface{})
+	for k, v := range s.s.m.Arguments {
+		mapInf[k] = v
+	}
+
+	// if not set then we return a default value based on the denoted type
+	v, err := dotnotation.Get(mapInf, pth)
+	if err != nil {
+		v, err = s.resolveDefault(pth, &arg)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// validate the data
+	if arg.Schema != nil {
+		if err := s.validateArg(pth, &arg, v); err != nil {
+			return nil, err
+		}
+	}
+
+	return v, nil
+}
+
+// resolveDefault resolves the default value of an argument from the manifest
+func (s *TplStencil) resolveDefault(pth string, arg *configuration.Argument) (interface{}, error) {
+	if arg.Default != nil {
+		return arg.Default, nil
+	}
+
+	if arg.Required {
+		return nil, fmt.Errorf("module %q requires argument %q but is not set", s.t.Module.Name, pth)
+	}
+
+	// json schema convention is to define "type" as the top level key.
+	typ, ok := arg.Schema["type"]
+	if !ok {
+		// fallback to the deprecated arg.Type
+		typ = arg.Type //nolint:staticcheck // Why: Compat
+
+		// If arg.Type isn't set then we have no type information
+		// so return nothing. This is likely problematic so a linter
+		// should warn on this.
+		if arg.Type == "" { //nolint:staticcheck // Why: Compat
+			return nil, nil
+		}
+	}
+	typs, ok := typ.(string)
+	if !ok {
+		return nil, fmt.Errorf("module %q argument %q has invalid type: %v", s.t.Module.Name, pth, typ)
+	}
+
+	var v interface{}
+	switch typs {
+	case "map", "object":
+		v = make(map[interface{}]interface{})
+	case "list", "array":
+		v = []interface{}{}
+	case "boolean", "bool":
+		v = false
+	case "integer", "int", "number":
+		v = 0
+	case "string":
+		v = ""
+	default:
+		return "", fmt.Errorf("module %q argument %q has invalid type %q", s.t.Module.Name, pth, typs)
+	}
+
+	return v, nil
+}
+
+// resolveFrom resoles the "from" field of an argument
+func (s *TplStencil) resolveFrom(ctx context.Context, pth string, arg *configuration.Argument) (*configuration.Argument, error) {
+	foundModuleInDeps := false
+	ourMf, err := s.t.Module.Manifest(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that the module imports the referenced module
+	for _, m := range ourMf.Modules {
+		if m.Name == arg.From {
+			foundModuleInDeps = true
+		}
+	}
+	if !foundModuleInDeps {
+		return nil, fmt.Errorf(
+			"module %q argument %q references an argument in module %q, but doesn't list it as a dependency",
+			s.t.Module.Name, pth, arg.From,
+		)
+	}
+
+	// Get the manifest for the referenced module
+	var fromMf *configuration.TemplateRepositoryManifest
```

Comment:
This might be a preemptive comment (you might've already done this) but I'd call out in the docs that from isn't transitive - it can only go one layer deep. I don't know why we'd ever end up there tho

Reply:
Good call